balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +854,268 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
+// causes pickfirst to reach the end of the address list and failures are
+// reported out of order.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		ret := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			ret.Reset(0)
+		}()
+		return ret
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	servers := newHangingServerGroup(t, 3)
+	rb := manual.NewBuilderWithScheme("whatever")
+	addrs := resolverAddrsFromHangingServers(servers)
+	rb.InitialState(resolver.State{Addresses: addrs})
+	cc, err := grpc.NewClient("whatever:///this-gets-overwritten",
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithResolvers(rb))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cc.Close()
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if err := servers[0].awaitContacted(ctx); err != nil {
+		t.Fatalf("Server with address %q not contacted: %v", addrs[0], err)
+	}
+
+	// Ensure no other servers are contacted.
+	if got, want := servers[1].isContacted(), false; got != want {
+		t.Fatalf("Server[%q].isContacted() = %t, want %t", addrs[1], got, want)
+	}
+	if got, want := servers[2].isContacted(), false; got != want {
+		t.Fatalf("Server[%q].isContacted() = %t, want %t", addrs[2], got, want)
+	}
+
+	// Make the happy eyeballs timer fire twice so that pickfirst reaches the
+	// last address in the list.
+	timerCh <- struct{}{}
+
+	// Verify that the second server is contacted and 3rd isn't.
+	if err := servers[1].awaitContacted(ctx); err != nil {
+		t.Fatalf("Server with address %q not contacted: %v", addrs[1], err)
+	}
+
+	if got, want := servers[2].isContacted(), false; got != want {
+		t.Fatalf("Server[%q].isContacted() = %t, want %t", addrs[2], got, want)
+	}
+	timerCh <- struct{}{}
+	if err := servers[2].awaitContacted(ctx); err != nil {
+		t.Fatalf("Server with address %q not contacted: %v", addrs[2], err)
+	}
+
+	// First SubConn Fails.
+	servers[0].closeConn()
+
+	// No TF should be reported until the first pass is complete.
+	shortCtx, shortCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer shortCancel()
+
+	testutils.AwaitNotState(shortCtx, t, cc, connectivity.TransientFailure)
+
+	// Move off the end of the list, pickfirst should still be waiting for TFs
+	// to be reported.
+	timerCh <- struct{}{}
```

Comment:
This is slightly off the gRFC. Not a major concern in terms of correctness. The gRFC does say that if we are starting a connection to the last address on the list, we shouldn't be starting the timer. But our code does start the timer.